src/common/erc721/ERC721Parser.ts

```diff
@@ -0,0 +1,597 @@
+        });
+
+        return Promise.all(promises)
+            .then(() => {
+                return Promise.resolve();
+            })
+            .catch((err: Error) => {
+                winston.error(`Error parsing blocks: ${blockNumbers}, error: ${err}`);
+            });
+    }
+
+    public async parse(block): Promise<any[]> {
+        winston.info(`ERC721Parser.parse(${block.number})`);
+
+        const transactions = this.extractTransactions(block);
+        const transactionIDs = this.getTransactionIDs(transactions);
+        const receipts = await this.fetchReceiptsFromTransactionIDs(transactionIDs);
+        const mergedTransactions = await this.attachReceiptsToTransactions(transactions, receipts);
+
+        const results = await this.updateTransactionsInDatabase(mergedTransactions);
+        winston.info(`ERC721Parser.updateTransactionsInDatabase, rows: ${results.length}, block: ${block.number}`);
+
+        const contractAddresses = await this.extractContractAddresses(transactions);
+        const contracts = await this.getERC721Contracts(contractAddresses);
+
+        const savedContracts = await this.updateERC721ContractsInDatabase(contracts);
+        if (savedContracts.length > 0) {
+            winston.info(`ERC721Parser.updateERC721ContractsInDatabase, rows: ${savedContracts.length}, block: ${block.number}`);
+        }
+
+        const transactionOperations = await this.parseTransactionOperations(transactions, savedContracts);
+
+        const savedTransactionOperations = await this.updateTransactionOperationsInDatabase(transactionOperations);
+        if (savedTransactionOperations.length > 0) {
+            winston.info(`ERC721Parser.updateTransactionOperationsInDatabase, rows: ${savedTransactionOperations.length}, block: ${block.number}`);
+        }
+
+        const savedTokenOwnerships = await this.updateTokenOwnership(block.number);
+        // TODO: print out the number of rows affected
+        // winston.info(`ERC721Parser.updateTokenOwnership, rows: ${savedTokenOwnerships}, block: ${block.number}`);
+
+        return Promise.resolve(savedTokenOwnerships);
+    }
+
+    public extractTransactions(block): any[] {
+        return this.getRawTransactions(block).map((tx: ITransaction) => {
+            return new ERC721Transaction(this.extractTransaction(block, tx));
+        });
+    }
+
+    public getTransactionIDs(transactions): string[] {
+        return transactions.map((tx: IExtractedTransaction) => tx._id);
+    }
+
+    public async fetchReceiptsFromTransactionIDs (transactionIDs: string[]) {
+        const batchLimit = 300
+        const chunk = (list, size) => list.reduce((r, v) =>
+            (!r.length || r[r.length - 1].length === size ?
+                r.push([v]) : r[r.length - 1].push(v)) && r
+            , []);
+        const chunkTransactions = chunk(transactionIDs, batchLimit)
+
+        try {
+            const receipts = await Bluebird.map(chunkTransactions, (chunk: any) => {
```

Comment:
can we store events for each transaction and just parse those then? otherwise you duplicate fetching transaction logic

Reply:
Yeah this is duplicate, will extract the code to a common file.